patches/server/0957-add-config-for-velocity-secret-key-file.patch

```diff
@@ -0,0 +1,96 @@
+index c4934979b1ed85bfc4f8d9e6f8848b2beaad95c3..6703468b0c0547502aee8c788d77a624c41d56e1 100644
+--- a/src/main/java/com/destroystokyo/paper/proxy/VelocityProxy.java
++++ b/src/main/java/com/destroystokyo/paper/proxy/VelocityProxy.java
+@@ -4,6 +4,7 @@ import io.papermc.paper.configuration.GlobalConfiguration;
+ import com.google.common.net.InetAddresses;
+ import com.mojang.authlib.GameProfile;
+ import com.mojang.authlib.properties.Property;
++import java.io.IOException;
+ import java.net.InetAddress;
+ import java.security.InvalidKeyException;
+ import java.security.MessageDigest;
+@@ -33,12 +34,12 @@ public class VelocityProxy {
+ 
+         try {
+             final Mac mac = Mac.getInstance("HmacSHA256");
+-            mac.init(new SecretKeySpec(GlobalConfiguration.get().proxies.velocity.secret.getBytes(java.nio.charset.StandardCharsets.UTF_8), "HmacSHA256"));
++            mac.init(new SecretKeySpec(GlobalConfiguration.get().proxies.velocity.getSecret().getBytes(java.nio.charset.StandardCharsets.UTF_8), "HmacSHA256"));
+             final byte[] mySignature = mac.doFinal(data);
+             if (!MessageDigest.isEqual(signature, mySignature)) {
+                 return false;
+             }
+-        } catch (final InvalidKeyException | NoSuchAlgorithmException e) {
++        } catch (final InvalidKeyException | IOException | NoSuchAlgorithmException e) {
+             throw new AssertionError(e);
+         }
+ 
+diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+index 8d442c5a498ecf288a0cc0c54889c6e2fda849ce..948b0751b355bdbaaa8d21133c11beade1fa19cf 100644
+--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
++++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+@@ -13,6 +13,9 @@ import org.spongepowered.configurate.objectmapping.meta.Comment;
+ import org.spongepowered.configurate.objectmapping.meta.Required;
+ import org.spongepowered.configurate.objectmapping.meta.Setting;
+ 
++import java.io.IOException;
++import java.nio.file.Files;
++import java.nio.file.Path;
+ import java.util.List;
+ import java.util.Map;
+ import java.util.Objects;
+@@ -84,6 +87,16 @@ public class GlobalConfiguration extends ConfigurationPart {
+             public boolean enabled = false;
+             public boolean onlineMode = false;
+             public String secret = "";
++            public String secretFile = "";
++            public String getSecret() throws IOException {
++                if (!this.secret.isEmpty()) {
++                    return this.secret;
++                } else if (!this.secretFile.isEmpty()) {
++                    return String.join("", Files.readAllLines(Path.of(this.secretFile)));
++                } else {
++                    return "";
++                }
++            }
```

Comment:
I'm not a fan of adding unneeded methods to the config class. Preferably it's just fields. Perhaps we should have a custom type to handle secret in text or secret in file?